File: dromozoa/compiler/stage1_prologue.mjs
// Copyright (C) 2022 Tomoyuki Fujimori <moyu@dromozoa.com>
//
// This file is part of dromozoa.
//
// dromozoa is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// dromozoa is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// Under Section 7 of GPL version 3, you are granted additional
// permissions described in the GCC Runtime Library Exception, version
// 3.1, as published by the Free Software Foundation.
//
// You should have received a copy of the GNU General Public License
// and a copy of the GCC Runtime Library Exception along with
// dromozoa.  If not, see <http://www.gnu.org/licenses/>.

import * as fs from "fs";

globalThis.fs = fs;

//-------------------------------------------------------------------------

const type_impl = {
  undefined: "nil",
  number: "number",
  string: "string",
  boolean: "boolean",
  symbol: "userdata",
  bigint: "userdata",
};

const type = v => {
  const t = type_impl[typeof v];
  if (t) {
    return t;
  } else if (v.LuaTable) {
    return "table";
  } else if (v.LuaFunction) {
    return "function";
  } else {
    return "userdata";
  }
};

//-------------------------------------------------------------------------

class LuaError extends Error {
  constructor(msg) {
    super(msg);
    this.name = "LuaError";
    this.msg = msg;
  }

  LuaError() {}
}

class LuaTable extends Map {
  constructor() {
    super();
    this.n = 0;
    this.metatable = undefined;
  }

  LuaTable() {}
}

const LuaFunction = f => {
  f.LuaFunction = true;
  return f;
};

//-------------------------------------------------------------------------

const OP_GETTABLE = (t, k) => {
  if (t.LuaTable) {
    const v = t.get(k);
    if (v !== undefined || !t.metatable) {
      return v;
    }
    const metafield = t.metatable.get("__index");
    if (metafield === undefined) {
      return undefined;
    }
    if (metafield.LuaTable) {
      return OP_GETTABLE(metafield, k);
    } else {
      return metafield(t, k)[0];
    }
  } else if (typeof t === "string") {
    return D.string_metatable.get("__index").get(k);
  } else {
    return t[k];
  }
};

const OP_SETLIST = (t, r) => {
  const n = r.length;
  for (let i = 1; i <= n; ++i) {
    const v = r[i - 1];
    if (v !== undefined) {
      t.set(i, v);
    } else {
      t.delete(i);
    }
  }
  if (t.n !== undefined && t.n < n) {
    t.n = n;
  }
};

const OP_LEN = t => {
  if (t.LuaTable) {
    if (t.n === undefined) {
      let i = 1;
      for (; t.get(i) !== undefined; ++i) {}
      t.n = i - 1;
    }
    return t.n;
  } else if (typeof t === "string") {
    return E.get("string").get("len")(t)[0];
  } else {
    return t.length;
  }
};

const OP_CHECKNUMBER = (v, msg) => {
  const t = typeof v;
  if (t === "number") {
    return v;
  } else if (t === "string") {
    v = +v;
    if (Number.isNaN(v)) {
      throw new LuaError(msg + " (number expected, got string)");
    }
    return v;
  } else {
    throw new LuaError(msg + " (number expected, got " + type(v) + ")");
  }
};

//-------------------------------------------------------------------------

const D = {
  string_metatable: new LuaTable(),

  getmetafield: (t, ev) => {
    if (t.LuaTable) {
      if (!t.metatable) {
        return undefined;
      }
      return t.metatable.get(ev);
    } else if (typeof t === "string") {
      return D.string_metatable.get(ev);
    }
  },

  native_new: (f, ...args) => {
    return new f(...args);
  },

  native_function: f => {
    return (...args) => {
      return f(...args)[0];
    };
  },

  native_call: (t, k, ...args) => {
    return t[k].apply(t, args);
  },

  array_pack: (...args) => {
    return args;
  },

  array_unpack: LuaFunction(r => {
    return r;
  }),

  array_from: (t, m, n) => {
    const r = [];
    for (let i = m; i <= n; ++i) {
      r[i - m] = OP_GETTABLE(t, i);
    }
    return r;
  },
};

//-------------------------------------------------------------------------

const E = new LuaTable();
E.set("globalThis", globalThis);
E.set("type", type);
E.set("dromozoa", D);

{
  const P = new LuaTable();
  P.set("preload", new LuaTable());
  P.set("loaded", new LuaTable());
  E.set("package", P);
}

{
  const A = new LuaTable();
  for (let i = 0; i < process.argv.length; ++i) {
    A.set(i - 1, process.argv[i]);
  }
  A.n = process.argv.length - 2;
  E.set("arg", A);
}
const A = process.argv.slice(2);

//-------------------------------------------------------------------------

E.set("error", msg => {
  throw new LuaError(msg);
});

E.set("getmetatable", t => {
  if (t.LuaTable) {
    if (!t.metatable) {
      return undefined;
    }
    const metafield = t.metatable.get("__metatable");
    if (metafield === undefined) {
      return t.metatable;
    }
    return metafield;
  } else if (typeof t === "string") {
    return D.string_metatable;
  }
});

E.set("setmetatable", (t, metatable) => {
  if (t.metatable && t.metatable.get("__metatable") !== undefined) {
    throw new LuaError("cannot change a protected metatable");
  }
  t.metatable = metatable;
  return t;
});

E.set("select", LuaFunction((k, ...args) => {
  if (k === "#") {
    return [ args.length ];
  } else {
    return args.slice(k - 1);
  }
}));

E.set("pcall", LuaFunction((f, ...args) => {
  try {
    return [ true, ...f(...args) ];
  } catch (e) {
    if (e.LuaError) {
      return [ false, e.msg ];
    } else {
      return [ false, e.toString() ];
    }
  }
}));

//-------------------------------------------------------------------------

const S0 = [];
const S1 = [undefined];

File: dromozoa/compiler/stage1_prologue.lua
return "import * as fs from \"fs\";\nglobalThis.fs = fs;\nconst type_impl = {\n  undefined: \"nil\",\n  number: \"number\",\n  string: \"string\",\n  boolean: \"boolean\",\n  symbol: \"userdata\",\n  bigint: \"userdata\",\n};\nconst type = v => {\n  const t = type_impl[typeof v];\n  if (t) {\n    return t;\n  } else if (v.LuaTable) {\n    return \"table\";\n  } else if (v.LuaFunction) {\n    return \"function\";\n  } else {\n    return \"userdata\";\n  }\n};\nclass LuaError extends Error {\n  constructor(msg) {\n    super(msg);\n    this.name = \"LuaError\";\n    this.msg = msg;\n  }\n  LuaError() {}\n}\nclass LuaTable extends Map {\n  constructor() {\n    super();\n    this.n = 0;\n    this.metatable = undefined;\n  }\n  LuaTable() {}\n}\nconst OP_NEWTABLE = () => {\n  return new LuaTable();\n};\nconst OP_NEWFUNCTION = f => {\n  f.LuaFunction = true;\n  return f;\n};\nconst OP_SETTABLE = (t, k, v) => {\n  if (t.LuaTable) {\n    if (t.get(k) === undefined && t.metatable) {\n      const metafield = t.metatable.get(\"__newindex\");\n      if (metafield !== undefined) {\n        if (metafield.LuaTable) {\n          OP_SETTABLE(metafield, k, v);\n        } else {\n          OP_CALL(metafield, [t, k, v]);\n        }\n        return;\n      }\n    }\n    if (v !== undefined) {\n      if (t.n !== undefined && Number.isInteger(k) && k > t.n && k !== ++t.n) {\n        t.n = undefined;\n      }\n      t.set(k, v);\n    } else {\n      if (t.n !== undefined && Number.isInteger(k) && k !== t.n--) {\n        t.n = undefined;\n      }\n      t.delete(k);\n    }\n  } else {\n    if (v !== undefined) {\n      t[k] = v;\n    } else {\n      delete t[k];\n    }\n  }\n};\nconst OP_GETTABLE = (t, k) => {\n  if (t.LuaTable) {\n    const v = t.get(k);\n    if (v !== undefined || !t.metatable) {\n      return v;\n    }\n    const metafield = t.metatable.get(\"__index\");\n    if (metafield === undefined) {\n      return undefined;\n    }\n    if (metafield.LuaTable) {\n      return OP_GETTABLE(metafield, k);\n    } else {\n      return OP_CALL(metafield, [t, k])[0];\n    }\n  } else if (typeof t === \"string\") {\n    return OP_GETTABLE(D.string_metatable.get(\"__index\"), k);\n  } else {\n    return t[k];\n  }\n};\nconst OP_CALL = (f, args) => {\n  if (f.LuaFunction) {\n    return f(...args);\n  } else if (f.LuaTable) {\n    return OP_CALL(f.metatable.get(\"__call\"), [f, ...args]);\n  } else {\n    return [ f(...args) ];\n  }\n};\nconst OP_SELF = (t, k, args) => {\n  const f = OP_GETTABLE(t, k);\n  if (f.LuaFunction) {\n    return f(t, ...args);\n  } else if (f.LuaTable) {\n    return OP_CALL(f.metatable.get(\"__call\"), [f, t, ...args]);\n  } else {\n    return [ f.apply(t, args) ];\n  }\n};\nconst OP_SETLIST = (t, r) => {\n  for (let i = 0; i < r.length; ++i) {\n    OP_SETTABLE(t, i + 1, r[i]);\n  }\n};\nconst OP_LEN = t => {\n  if (t.LuaTable) {\n    if (t.n === undefined) {\n      let i = 1;\n      for (; t.get(i) !== undefined; ++i) {}\n      t.n = i - 1;\n    }\n    return t.n;\n  } else if (typeof t === \"string\") {\n    return D.string_len(t);\n  } else {\n    return t.length;\n  }\n};\nconst OP_CLOSE = t => {\n  if (t !== undefined) {\n    OP_CALL(t.metatable.get(\"__close\"), [t]);\n  }\n};\nconst OP_CHECKNUMBER = (v, msg) => {\n  const t = typeof v;\n  if (t === \"number\") {\n    return v;\n  } else if (t === \"string\") {\n    v = +v;\n    if (Number.isNaN(v)) {\n      throw new LuaError(msg + \" (number expected, got string)\");\n    }\n    return v;\n  } else {\n    throw new LuaError(msg + \" (number expected, got \" + type(v) + \")\");\n  }\n};\nconst D = {\n  string_metatable: OP_NEWTABLE(),\n  arg: [],\n  getmetafield: (t, ev) => {\n    if (t.LuaTable) {\n      if (!t.metatable) {\n        return undefined;\n      }\n      return t.metatable.get(ev);\n    } else if (typeof t === \"string\") {\n      return D.string_metatable.get(ev);\n    }\n  },\n  native_new: (f, ...args) => {\n    return new f(...args);\n  },\n  native_function: f => {\n    return (...args) => {\n      return f(...args)[0];\n    };\n  },\n  native_call: (t, k, ...args) => {\n    return t[k].apply(t, args);\n  },\n  array_pack: (...args) => {\n    return args;\n  },\n  array_unpack: OP_NEWFUNCTION(r => {\n    return r;\n  }),\n  array_from: (t, m, n) => {\n    const r = [];\n    for (let i = m; i <= n; ++i) {\n      r[i - m] = OP_GETTABLE(t, i);\n    }\n    return r;\n  },\n};\nconst E = OP_NEWTABLE();\nOP_SETTABLE(E, \"globalThis\", globalThis);\nOP_SETTABLE(E, \"type\", type);\nOP_SETTABLE(E, \"dromozoa\", D);\nconst P = OP_NEWTABLE();\nOP_SETTABLE(P, \"preload\", OP_NEWTABLE());\nOP_SETTABLE(P, \"loaded\", OP_NEWTABLE());\nOP_SETTABLE(E, \"package\", P);\nOP_SETTABLE(E, \"error\", msg => {\n  throw new LuaError(msg);\n});\nOP_SETTABLE(E, \"getmetatable\", t => {\n  if (t.LuaTable) {\n    if (!t.metatable) {\n      return undefined;\n    }\n    const metafield = t.metatable.get(\"__metatable\");\n    if (metafield !== undefined) {\n      return metafield;\n    }\n    return t.metatable;\n  } else if (typeof t === \"string\") {\n    return D.string_metatable;\n  }\n});\nOP_SETTABLE(E, \"setmetatable\", (t, metatable) => {\n  if (t.metatable && t.metatable.get(\"__metatable\") !== undefined) {\n    throw new LuaError(\"cannot change a protected metatable\");\n  }\n  t.metatable = metatable;\n  return t;\n});\nOP_SETTABLE(E, \"select\", OP_NEWFUNCTION((k, ...args) => {\n  if (k === \"#\") {\n    return [ args.length ];\n  } else {\n    return args.slice(k - 1);\n  }\n}));\nOP_SETTABLE(E, \"pcall\", OP_NEWFUNCTION((f, ...args) => {\n  try {\n    return [ true, ...OP_CALL(f, args) ];\n  } catch (e) {\n    if (e.LuaError) {\n      return [ false, e.msg ];\n    } else {\n      return [ false, e.toString() ];\n    }\n  }\n}));\n"
